internal/server/recon/recognize.go

```diff
@@ -73,6 +77,189 @@ func RecognizePlaces(
+			for _, id := range entity.GetOutIds() {
+				// TODO: add types to the response
+				places = append(places, &pb.RecognizePlacesResponse_Place{Dcid: id})
+			}
+			for span := range id2spans[entity.GetInId()] {
+				span2item[span] = &pb.RecognizePlacesResponse_Item{Span: span, Places: places}
+			}
+		}
+
+		// get the list of resolved spans and sort them by more words first. If two
+		// spans have the same number of words, sort alphabetically.
+		spans := []string{}
+		for span := range span2item {
+			spans = append(spans, span)
+		}
+		sort.Slice(spans, func(i, j int) bool {
+			iCount := strings.Count(spans[i], " ")
+			jCount := strings.Count(spans[j], " ")
+			if iCount > jCount {
+				return true
+			} else if jCount > iCount {
+				return false
+			} else {
+				return spans[i] < spans[j]
+			}
+		})
+
+		// Get the response items
+		queryRespItems := getItemsForSpans(spans, query, span2item)
+		resp.QueryItems[query] = &pb.RecognizePlacesResponse_Items{Items: queryRespItems}
+	}
+	return resp, nil
+}
+
+// Takes a query and list of spans and gets the corresponding list of recognize
+// places response items. With the list of spans, we want the response items to
+// have a subset of non-overlapping spans. We do that with a greedy approach of
+// matching the longest span in query, getting the remaining query parts, and
+// recursively doing the match.
+func getItemsForSpans(spans []string, query string, span2item map[string]*pb.RecognizePlacesResponse_Item) []*pb.RecognizePlacesResponse_Item {
+	queryRespItems := []*pb.RecognizePlacesResponse_Item{}
+	// If empty query, return empty list
+	if len(query) == 0 {
+		return queryRespItems
+	}
+	// If empty list of spans, return the query as the only item
+	if len(spans) == 0 {
+		return append(queryRespItems, &pb.RecognizePlacesResponse_Item{Span: query})
+	}
+	span := spans[0]
+	queryParts := splitQueryBySpan(query, span)
+	for _, part := range queryParts {
+		if part == span {
+			queryRespItems = append(queryRespItems, span2item[span])
+		} else {
+			queryRespItems = append(queryRespItems, getItemsForSpans(spans[1:], part, span2item)...)
+		}
+	}
+	return queryRespItems
+}
+
+// Splits a query by a span into a list of parts like: non-span part, span part,
+// non-span part, span part, etc. If the query does not contain the span,
+// returns a list with the query as the only item.
+// Examples
+// query: "ab cd ef ab g", span: "ab", result: ["ab", "cd ef", "ab", "g"]
+// query: "ab cd ef g", span: "jk", result: ["ab cd ef g"]
+func splitQueryBySpan(query string, span string) []string {
+	splits := strings.Split(query, span)
+	// if span is not found in query, return just the query
+	if len(splits) < 2 {
+		return []string{query}
+	}
+	queryParts := []string{}
+	// prefix to add to the next split to be processed
+	nextPartPrefix := ""
+	// go through all the splits and add the span between each split if it was a
+	// valid split (we only want to split on complete words)
```

Comment:
I wondered if this can be made more readable, and it slightly felt like split was adding complexity.  Here's a version that reads better, but I'm not sure because I wrote it :).  Anyway, see if that seems better or you'd prefer to stick with split + cleaning.

```
needle = 'foo ba'
haystack = 'foo bar foo ba r'

i = 0
parts = ['']
while i < len(haystack):
  if i == 0 or haystack[i - 1] == ' ':
    # A match starts at beginning of string or after a space!

    if haystack[i:].startswith(needle):
      eidx = i + len(needle)

      if len(haystack[i:]) == len(needle) or haystack[eidx] in [',', '.', ' ']:
        # A match ends with a space, comma, full-stop or end-of-string!
        parts.append(haystack[i:eidx])

        # Add an empty string for the next span.
        parts.append('')

        i += len(needle)
        continue

  # No luck, append to the last span
  parts[len(parts) - 1] += haystack[i]
  i += 1

# TODO: trim empty entries
```

Reply:
i like this, updated